osu.Game/Screens/Select/PlaySongSelect.cs

```diff
@@ -22,29 +22,40 @@
 
 namespace osu.Game.Screens.Select
 {
-    public class PlaySongSelect : SongSelect
+    public class PlaySongSelect : SongSelect, ILeaderboardScoreSource
```

Comment:
Is this still required? I find this implementation exceedingly weird. It's forwarding the score source to `playBeatmapDetailArea.Leaderboard.Scores`, but `BeatmapLeaderboard` is already an `ILeaderboardScoreSource`?

And, part of the bigger picture, I am very confused about this `ILeaderboardScoreSource` stuff. Why is a leaderboard a score source and/or why is the leaderboard not the _single_ score source ever? It's like, `SoloGameplayLeaderboard` binds _outwards_ now but still has a perfectly accessible `.Add()` method.

The beatmap leaderboard pulls things from the database or online, the multiplayer leaderboard opens spectator channels and populates its own scores/manages their own states, why can't solo player do this too?

Reply:
Is it just that `PlaySongSelect` is an `ILeaderboardScoreSource` which is cached for when entering `Player` so that the leaderboard can then look it up?

It's just too much. The only reason that I can come up with this is to save on API requests? Which is fine, but in the process of doing that the code quality is completely destroyed and logic now goes all around the place and is unreadable.

Reply:
It's just for the caching, yes.

It means that whatever is displayed at song select is used during gameplay (similar to stable).

The whole flow will need to change in the future (as I want to be able to dynamically load in/out scores using ongoing web requests), but the idea was to keep things simple in this implementation.

Reply:
I don't feel good merging this. I suggest refactoring it in this PR and/or a pre-requisite PR that has a saner structure.

Reply:
I'll see what I can do.

Would you prefer something like a bindable exposed on `Player` which is bound to by `PlaySongSelect`? Is the main issue the DI interface?

Reply:
The main issue I have is the DI interface, yes.

However it's a question of what you're actually trying to achieve here:
- Is it about retrieval type (i.e. which tab is selected in song select) and if so why can't that be re-queried again? As you mentioned, it's going to need some extra logic anyway to re-query the intermediate scores as the player progresses up the rankings so there's little point (in my mind) in taking on song select's rankings.  
  - The selected tab is already stored to the `OsuConfigManager` (`BeatmapDetailTab`+`BeatmapDetailModsFilter`) so there shouldn't be any problems.
- Is it about saving on web requests? If so, perhaps a separate component is better fit for this? For example, what happens when a user replays a map, is the client going to re-query all the intermediate scores for the second play? If it does, is there any point on saving web requests - I assume we want the most recent data (especially for newly-ranked beatmaps).

Reply:
If you look back at the history of this PR, the component was made in a way that would always theoretically work with `PlaylistPlayer`. As long as a parent has a score source, it would just-work.

I removed the playlist portion because frenzi made the incorrect assumption about `AggregateScore`s being per-beatmap.

> * Is it about retrieval type (i.e. which tab is selected in song select) and if so why can't that be re-queried again?

Yes, things like the mods filter etc. would need to be communicated if you wanted to do a second redundant lookup.

>   * The selected tab is already stored to the `OsuConfigManager` (`BeatmapDetailTab`+`BeatmapDetailModsFilter`) so there shouldn't be any problems.

> * Is it about saving on web requests?

Yeah, I wouldn't want to query the same thing twice at two consecutive screens, for sure.

Reply:
Well `PlaylistPlayer` is also a child of `Playlists` (apparently what it's called, no `Screen` suffix...) and _not_ the `PlaylistsRoomSubscren`. So I don't believe it would work there unless I'm missing something.

Are you suggesting that `Playlists` implement this new interface, and then the subscreen uses scores DI'd from that? And then `Playlists` somehow knows about all the interactions going on such as what if the leaderboard inside the subscreen is scrolled - should more scores come up, etc.

Reply:
The playlist implementation was equally weird with redirection, yes. I'm not saying that would return (currently playlist only retrieves scores when entering the results screen) – it needs more thought for sure.

I do think there's value in having a shared data source between song select and player though (in terms of streamlining requests and realm overhead), rather than requesting at each location. I'll give a bindable source a try and see how things go.

Reply:
I've changed the flow, see if it works any better to your eye.